File: solutions/php/pascals-triangle/1/PascalsTriangle.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

function pascalsTriangleRows(int $rowCount): array
{
    // Handle edge case
    if ($rowCount <= 0) {
        return [];
    }
    
    $triangle = [];
    
    for ($row = 0; $row < $rowCount; $row++) {
        $currentRow = [];
        
        // Calculate each element in the current row
        for ($col = 0; $col <= $row; $col++) {
            // Use binomial coefficient formula: C(row, col) = row! / (col! * (row-col)!)
            // But we can calculate it iteratively to avoid large factorials
            
            // For the first and last elements in a row, the value is always 1
            if ($col === 0 || $col === $row) {
                $currentRow[] = 1;
            } else {
                // Use Pascal's rule: C(n,k) = C(n-1,k-1) + C(n-1,k)
                // We need to get values from the previous row
                $prevRow = $triangle[$row - 1];
                $value = $prevRow[$col - 1] + $prevRow[$col];
                $currentRow[] = $value;
            }
        }
        
        $triangle[] = $currentRow;
    }
    
    return $triangle;
}

File: solutions/vbnet/pascals-triangle/1/PascalsTriangle.vb
Imports System.Collections.Generic

Public Module PascalsTriangle
    Public Function Calculate(ByVal rows As Integer) As IEnumerable(Of IEnumerable(Of Integer))
        Dim triangle As New List(Of List(Of Integer))()

        ' Handle the edge case where 0 rows are requested
        If rows <= 0 Then
            Return triangle
        End If

        For i As Integer = 0 To rows - 1
            Dim row As New List(Of Integer)()
            
            ' The first element of every row is always 1
            row.Add(1)

            ' For rows after the first, calculate the inner elements
            If i > 0 Then
                Dim prevRow As List(Of Integer) = triangle(i - 1)
                
                ' Sum the two elements directly above from the previous row
                For j As Integer = 1 To i - 1
                    row.Add(prevRow(j - 1) + prevRow(j))
                Next
                
                ' The last element of every row (except the first) is always 1
                row.Add(1)
            End If

            triangle.Add(row)
        Next

        Return triangle
    End Function
End Module